### ISSUE TITLE
Implement a C#-version of the Go-implementation

### ISSUE THREAD

**Issue by HenrikWM** [HenrikWM/anonymous-tokens#2]

**Comment by HenrikWM** [comment]
C# pseudo-code from [Martin Strand](https://twitter.com/martinstrand) will be the starting point and to be implemented with the [.Net Core 3.1 portable Bouncy Castle NuGet-package](https://www.nuget.org/packages/Portable.BouncyCastle):

```
multiplier:
protected virtual ECMultiplier multiplier()
{
    return new FixedPointCombMultiplier();
}
// Poenget med denne er å velge hvilken algoritme man skal bruke for å utføre punktmultiplikasjon. FixedPointCombMultiplier er antageligvis helt grei.


// Setup, kjøres på alle steder
Setup
  // P-256 ser ut til å ha en fin implementasjon, mens curve25519 visstnok er experimental. Hvis ikke hadde curve25519 vært førstevalget
  ECCurve curve = ECNamedCurveTable.GetByName("secp256k1");


// Backend, kjøres én gang per nøkkel. Privat nøkkel distribueres til
// smittestopp-backend og verifiseringsløsningen. Offentlig nøkkel
// publiseres på egnet vis. Enkel løsning: Lagre i appen.
KeyGen(ECCurve curve, ECPoint gen)
  // Velg k tilfeldig, kodet som
  BigInteger k = <random mellom 0 og curve.FiniteField.getCharacteristic>
  ECPoint K = multiplier(gen, k) // Aner ikke helt hvordan notasjonen til dette skal være.
  return k, K // private, public


// Appen, kjøres i forbindelse med innlogging til idporten
// t og r lagres på dingsen, P sendes til idporten
Initiate(ECCurve curve)
  BigInteger r = <random mellom 0 og curve.FiniteField.getCharacteristic>
  t kan være tilfeldige bytes, de skal aldri regnes på
  ECPoint T = HashToCurve(t) // Denne må jeg se mer på
  ECPoint P = multiplier(T, r)

  // Sanity check på dette tidspunktet, for å sjekke at det vi gjør gir mening. Skal ikke med i ferdig kode:
  ECFieldElement x = ECCurve.FromBigInteger(r)
  ECFieldElement xi = x.Invert()
  ri = xi.ToBigInteger()
  assert: multiplier(P, ri) == T
  // Det hadde vært fint å slippe å konvertere mellom BigInteger og ECFieldElement. Vet ikke om det funker å bruke sistnevnte direkte.

  return t, r, P


// Kjøres på verifikasjonsserveren
GenerateToken(ECPoint P, BigInteger k)
  Q = ECMultiply(P, k)
  // Så må vi lage et ZK-bevis. Det tar vi når vi har fått resten her til å fungere


// Kjøres på appen
RandomiseToken(ECPoint Q, BigInteger r)
  // VerifyProof. Det tar vi senere.

  ECFieldElement x = ECCurve.FromBigInteger(r)
  ECFieldElement xi = x.Invert()
  ri = xi.ToBigInteger()
  W = multiplier(Q, ri)

  return W

// Kjøres på smittestopp-backend
VerifyToken(BigInteger k, BigInteger t, ECPoint W)
  ECPoint T = HashToCurve(t) // Denne må jeg se mer på
  ECPoint V = multiplier(T, k)
  sjekk: Er V == W?
```

**Comment by tjesi** [comment]
There is a lack of documentation of the Bounty Castle C# EC library, but the Java version of the same library provides some information: [Package org.bouncycastle.math.ec](http://javadox.com/org.bouncycastle/bcprov-jdk15on/1.53/org/bouncycastle/math/ec/package-summary.html).

**Comment by martstr** [comment]
```
HashToCurve(ECCurve curve, bytes t)
  hash = sha256(t) // SHA 256 bør være en ganske standard sak i C#
  x = tolk hash som et FieldElement, i range 0 < x < ECCurve.FiniteField.order
  // Dette er i prinisppet enkelt, siden ECCurve.FiniteField.order for kurven P-256 er et 256 bit tall. Imidlertid, dersom x > ECCurve.FiniteField.order hadde det vært fristende å bare kjøre en mod-operasjon for å få x liten nok. Det vil gjøre at lave x blir litt mer sannsynlige enn høye x, og derfor et sikkerhetsproblem. Da er det bedre å returnere med feil, og be om ny tilfeldig t. OK, la oss anta at alt er i orden hittil.
  y2 = x^3 + ECCurve.a*x + ECCurve.b // Alt dette skal være FieldElement, så forhåpentligvis gjør den modulo automatisk
  y = y2.sqrt() // Denne har 50 % sjanse for å lykkes. Hvis den ikke gjør det, be om ny tilfeldig t.
  
  T = ECCurve.CreatePoint(x, y)
  return T
```

**Comment by martstr** [comment]
```
CreateChallenge(ECPoint basePoint1, ECPoint basePoint2, ECPoint newPoint1, ECPoint newPoint2, ECPoint commitment1, ECPoint commitment2)
	for hvert Point i argumentlista:
		dytt Point.getEncoded() inn i SHA256 (evnt. lang bytearray). I stedet for getEncoded kan man kanskje også bruke getAffineXCoord() og getAffineYCoord() sammen. Poenget er bare at man må binde seg til noe helt unikt ved punktet.
	beregn hashen av hele greia
	return output fra hele SHA256 som en BigInteger, .Mod(basePoint1.Curve.Order)


CreateProof(ecParameters, BigInteger privateKey, ECPoint publicKey, ECPoint P, ECPoint Q)
	var r = <tilfeldig tall på samme måte som vi valgte r for brukeren. den rutinen kan antageligvis samles på ett sted>
	var X = ecParameters.G.Multiply(r);
	var Y = P.Multiply(r);
	BigInteger c = CreateChallenge(ecParameters.G, P, publicKey, Q, X, Y);

	BigInteger z = r.Subtract(c.Multiply(privateKey.D)).Mod(ecParamters.Curve.Order); // Troooor denne kan funke for å beregne z = r - c*privateKey
	return (c, z)

VerifyProof(ecParameters, ECPoint publicKey, ECPoint P, ECPoint Q, BigInteger c, BigInteger z)
	ECPoint X = ecParameters.G.Multiply(z).Add(publicKey.Multiply(c));
	ECPoint Y = P.Multiply(z).Add(Q.Multiply(c));
	cPrime = CreateChallenge(ecParameters.G, P, publicKey, Q, X, Y);
	return c == cPrime // Hvis de to er like, så er vi fornøyd. Hvis de er like når de *skal* være like, så kan vi virkelig være godt fornøyde.
```

**Comment by tjesi** [comment]
`public static ECPoint HashToCurve(ECCurve curve, byte[] t)
        {
            ECFieldElement temp, x, ax, x3, y, y2;
            var P = curve.Field.Characteristic;
            var sha256 = SHA256.Create();
            var hash = new BigInteger(sha256.ComputeHash(t));
            if (hash.CompareTo(BigInteger.One) < 0 || hash.CompareTo(P) >= 0)
                return null;
            x = curve.FromBigInteger(hash);     // x
            ax = x.Multiply(curve.A);           // Ax
            temp = x.Multiply(x);               // x^2
            x3 = temp.Multiply(x);              // x^3
            temp = x3.Add(ax);                  // x^3 + Ax
            y2 = temp.Add(curve.B);             // y^2 = x^3 + Ax + B
            y = y2.Sqrt();                      // y = sqrt(x^3 + Ax + B)
            if (y == null)
                return null;
            ECPoint T = curve.CreatePoint(x.ToBigInteger(), y.ToBigInteger());
            return T;
        }`

**Comment by tjesi** [comment]
```
public static (byte[] t, BigInteger r, ECPoint P) Initiate(ECCurve curve)
        {
            var random = new SecureRandom();
            BigInteger N = curve.Order;
            BigInteger r;
            // Sample random 0 < r < N
            for (; ; )
            {
                r = new BigInteger(N.BitLength, random);
                if (r.CompareTo(BigInteger.One) < 0 || r.CompareTo(N) >= 0)
                    continue;
                break;
            }
            // Sample random bytes t such that x = hash(t) is a valid
            // x-coordinate on the curve. Then T = HashToCurve(t).
            var t = new byte[32];
            ECPoint T;
            for (; ; )
            {
                random.NextBytes(t);
                T = HashToCurve(curve, t);
                if (T == null)
                    continue;
                break;
            }
            // Compute P = r*T
            ECPoint P = T.Multiply(r);
            return (t, r, P);
        }
```

**Comment by HenrikWM** [comment]
Fixed in #4 #5 #6 🥇
